extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudStreamExtractor.java

```diff
@@ -181,62 +186,144 @@ public String getHlsUrl() {
+            final String url = transcodingJsonObject.getString("url");
+            final MediaFormat mediaFormat;
+            final int bitrate;
+
+            if (!isNullOrEmpty(url)) {
+                if (preset.contains("mp3")) {
+                    // Don't add the MP3 HLS stream if there is a progressive stream present
+                    // because the two have the same bitrate
+                    if (mp3ProgressiveInStreams && protocol.equals("hls")) {
+                        continue;
                     }
+                    mediaFormat = MediaFormat.MP3;
+                    bitrate = 128;
+                } else if (preset.contains("opus")) {
+                    mediaFormat = MediaFormat.OPUS;
+                    bitrate = 64;
+                } else {
+                    // Unknown format
+                    continue;
                 }
+
+                try {
+                    mediaUrl = getTranscodingUrl(url, protocol);
+                } catch (final Exception e) {
+                    // something went wrong when parsing this transcoding
+                    continue;
+                }
+                audioStreams.add(new AudioStream(mediaUrl, mediaFormat, bitrate));
             }
+        }
+    }
 
-        } catch (NullPointerException e) {
-            throw new ExtractionException("Could not get SoundCloud's track audio url", e);
+    private final static Pattern PATTERN_HTTPS_URLS_IN_HLS_MANIFESTS = Pattern.compile("((https?):((//)|(\\\\))+[\\w\\d:#@%/;$()~_?+-=\\\\.&]*)");
+
+    /** Parses a SoundCloud HLS manifest to get a single URL of HLS streams.
+     * <p>
+     * This method downloads the provided manifest URL, find all web occurrences using a regex, get
+     * the last segment URL, changes its segment range to {@code 0/track-length} and return this
+     * string.
+     * @param  hlsManifestUrl the URL of the manifest to be parsed
+     * @return a single URL that contains a range equal to the length of the track
+     */
+    private static String getSingleUrlFromHlsManifest(final String hlsManifestUrl) throws ParsingException {
+        final Downloader dl = NewPipe.getDownloader();
+        final String hlsManifestResponse;
+
+        try {
+            hlsManifestResponse = dl.get(hlsManifestUrl).responseBody();
+        } catch (final IOException | ReCaptchaException e) {
+            throw new ParsingException("Could not get SoundCloud HLS manifest");
         }
 
-        if (audioStreams.isEmpty()) {
-            throw new ContentNotSupportedException("HLS audio streams are not yet supported");
+        final List<String> hlsRangesList = new ArrayList<>();
+        final Matcher pattern_matches = PATTERN_HTTPS_URLS_IN_HLS_MANIFESTS
+                .matcher(hlsManifestResponse);
+
+        while (pattern_matches.find()) {
+            hlsRangesList.add(hlsManifestResponse.substring(pattern_matches.start(0),
+                    pattern_matches.end(0)));
```

Comment:
For context, [this is a SoundCloud HLS Manifest](https://gist.github.com/TobiGr/eabc45d311a887b3814882c7c25ecbe9).
From [RFC8216](https://tools.ietf.org/html/rfc8216#page-10), the HTTP Live Streaming standard:

> Lines in a Playlist file are terminated by either a single line feed
   character or a carriage return character followed by a line feed
   character.  Each line is a URI, is blank, or starts with the
   character '#'.  Blank lines are ignored.  Whitespace MUST NOT be
   present, except for elements in which it is explicitly specified.

This means, we do not need that big regex, but only something really small and fest to check if the line does not start with `#` and is not empty to know that we have a URL in that line. That should give a big performance improvement, especially for longer tracks. 

e.g. something like this (not tested, just typed into this review)
```Java
final String[] lines = hlsManifestResponse.split("\\r?\\n");
for (final String line : lines) {
    if (line.trim().length() != 0 && !line.startsWith("#") {
        hlsRangesList.add(line);
    }
}
```

@XiangRongLin @TiA4f8R Do you think that this is a good approach?

Reply:
@TobiGr The last line of the manifest is `#EXT-X-ENDLIST`, we can get the second last line, which is the last segment of the track and contains the range of the song, can't we?

Here is what it can looks:

```Java
        final String[] lines = hlsManifestResponse.split("\\r?\\n");
        final String hlsLastRangeUrl = hlsRangesList.get(lines.size() - 2);
        final String[] hlsLastRangeUrlArray = hlsLastRangeUrl.split("/");

        return HTTPS + hlsLastRangeUrlArray[2] + "/media/0/" + hlsLastRangeUrlArray[5] + "/"
                + hlsLastRangeUrlArray[6];
```

What do you think?

Reply:
The `#EXT-X-ENDLIST` is not required. We do not know for sure that the last URL is in the second last line. That's why I'd go through the the array reverse. You can start at the last index and go decrease it.
``` Java
for (int l = lines.length - 1; l >= 0; l--;) {
    if (line.trim().length() != 0 && !line.startsWith("#") {
        // you could also check , whether the line contains a valid URL
        // do what you wand to do here,
    }
}
```

Reply:
And for the first URL found, can I break the loop because it should be the last segment URL?

Reply:
Yes, you cloud do that. I'd do that to increase readbillity and also make error detection easier.
``` Java
for (int l = lines.length - 1; l >= 0; l--;) {
    final String line = lines[l];
    // get the last URL from manifest, because [add more info here]
    if (line.trim().length() != 0 && !lines.startsWith("#") {
        // you could also check , whether the line contains a valid URL
        final String[] hlsLastRangeUrlArray = line.split("/");
        return HTTPS + hlsLastRangeUrlArray[2] + "/media/0/" + hlsLastRangeUrlArray[5] + "/"
                + hlsLastRangeUrlArray[6];
    }
}
throw new ParsingException("Could not get any URL from HLS manifest");
```

Reply:
Your code perfectly works, I ran tests and URLs of HLS transcodings works and are equals to the track length!